conformance/test-server/src/main.rs

```diff
@@ -0,0 +1,298 @@
+use std::net::SocketAddr;
+
+use anyhow::Result;
+use clap::Parser;
+use futures::future;
+use hickory_proto::op::Message;
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::{TcpListener, UdpSocket},
+    task::JoinHandle,
+};
+
+mod handlers;
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let args = Args::parse();
+    let transport = args.transport;
+    let handler = match &args.handler[..] {
+        "base" => handlers::base_handler,
+        _ => {
+            return Err(anyhow::Error::msg("unknown handler"));
+        }
+    };
+
+    let mut handles = vec![];
+    if transport == "tcp" || transport == "both" {
+        let (_, tcp_handle) = tcp_server(([0, 0, 0, 0], args.port).into(), handler).await?;
+        handles.push(tcp_handle);
+    }
+    if transport == "udp" || transport == "both" {
+        let (_, udp_handle) = udp_server(([0, 0, 0, 0], args.port).into(), handler).await?;
+        handles.push(udp_handle);
+    }
+    let _ = future::join_all(handles).await;
+
+    Ok(())
+}
+
+#[derive(Parser)]
+#[clap(name = "Message responder test DNS server")]
+struct Args {
+    #[clap(default_value = "base", long = "handler")]
+    handler: String,
+    #[clap(default_value = "53", long = "port")]
+    port: u16,
+    #[clap(default_value = "both", long = "transport")]
+    transport: String,
+}
+
+async fn udp_server(
+    bind_addr: SocketAddr,
+    handler: HandlerFn,
+) -> Result<(SocketAddr, JoinHandle<()>)> {
+    let udp = UdpSocket::bind(bind_addr).await?;
+    let addr = udp.local_addr()?;
+    let handle = tokio::task::spawn(async move {
+        loop {
+            let mut read_buf = [0u8; 4096];
+            let to = match udp.recv_from(&mut read_buf).await {
+                Ok((_, to)) => to,
+                Err(e) => {
+                    println!("read error: {e:?}");
+                    continue;
+                }
+            };
+
+            println!(
+                "client request from udp/{}: {:?}",
+                to,
+                Message::from_vec(&read_buf)
+            );
+
+            match handler(&read_buf, Transport::Udp) {
+                Ok(Some(resp)) => {
+                    println!(
+                        "handler response to udp/{}: {:?}",
+                        to,
+                        Message::from_vec(&resp),
+                    );
+                    if let Err(e) = udp.send_to(&resp[..], to).await {
+                        println!("error sending response: {e:?}");
+                    }
+                }
+                Ok(None) => {
+                    println!("no response returned from handler; aborting");
+                }
+                Err(e) => {
+                    println!("error {e:?} returned from read handler; aborting");
+                }
+            };
+        }
+    });
+
+    Ok((addr, handle))
+}
+
+async fn tcp_server(
```

Comment:
It might still make sense to reuse `proto::TcpStream` for this, to avoid duplicating a bunch of logic here (which then has potential to get out of sync)?